### PR TITLE
feat/multicall

### PR DESCRIPTION
**Pull request by banteg**
### What I did

Added multicall context manager I initially outlined [here](https://github.com/eth-brownie/brownie/pull/1033#issuecomment-813139191).

Related issue: #1011 

### How I did it

I used [Multicall2](https://github.com/makerdao/multicall/pull/18#issuecomment-829313879) and `wrapt` library to replace results with proxies which are updated with actual values as soon as they are available.

Special thanks to @nymmrx for help with value wrappers.

### How to verify it

Here is a complex example:
```python
from brownie import *
import pandas as pd

def curve():
    curve_registry = Contract("0x7D86446dDb609eD0F5f8684AcF30380a356b2B4c")
    voter_proxy = Contract("0xF147b8125d2ef93FB6965Db97D6746952a133934")
    gauge_controller = Contract("0x2F50D538606Fa9EDD2B11E2446BEb18C9D5846bB")
    
    pool_count = curve_registry.pool_count()  # call 1
    with multicall() as caller:  # call 2
        pools = [caller(curve_registry).pool_list(i) for i in range(pool_count)]
    
    with multicall() as caller:  # call 3
        gauges = [caller(curve_registry).get_gauges(pool) for pool in pools]
    
    pools = [
        {'pool': Contract(pool), 'gauge': Contract(gauge[0][0])}
        for pool, gauge in zip(pools, gauges) if gauge[0][0] != ZERO_ADDRESS
    ]
    
    with multicall() as caller:  # call 4
        for pool in pools:
            pool['last_vote'] = caller(gauge_controller).last_user_vote(voter_proxy, pool['gauge'])
            pool['last_weight'] = caller(gauge_controller).vote_user_slopes(voter_proxy, pool['gauge'])
            pool['balance'] = caller(pool['gauge']).balanceOf(voter_proxy)
            pool['working_balance'] = caller(pool['gauge']).working_balances(voter_proxy)
            pool['virtual_price'] = caller(pool['pool']).get_virtual_price()

    for pool in pools:
        pool['last_weight'] = pool['last_weight'].dict()['power']
        pool['boost'] = pool['working_balance'] / pool['balance'] * 2.5 if pool['balance'] > 0 else 0
        pool['balance'] /= 1e18  # NOTE: all curve LP tokens have 18 decimals
        pool['virtual_price'] /= 1e18

    print(pd.DataFrame(pools)[['last_vote', 'last_weight', 'balance', 'boost', 'virtual_price']])
```

### Checklist

- [x] I have confirmed that my PR passes all linting checks
- [x] I have included test cases
- [x] I have updated the documentation
- [x] I have added an entry to the changelog

### CODE REVIEW

File: brownie/network/multicall.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+@contextmanager
+def multicall(block_identifier=None):
+    """
+    with multicall() as caller:
+        response = caller(contract).func(args)
+    """
+    queue = []
+    yield partial(Caller, queue)
+
+    multicall2 = Contract("0x5BA1e12693Dc8F9c48aAD8770482f4739bEeD696")
```
Comment (omarish) [inline brownie/network/multicall.py:52]: Suggestion: make this a default kwarg for multicall so it's not hard-coded in the framework?

```
@contextmanager
def multicall(block_identifier=None, multicall2_address="0x5BA1e12693Dc8F9c48aAD8770482f4739bEeD696"):
    ...
    multicall2 = Contract(multicall2_address)
```

Comment (banteg) [reply brownie/network/multicall.py:52]: I was thinking to make it even more flexible:
- live network => use `eth_call` code override to make it available pre-deploy
- dev network => seamlessly deploy a contract and cache it

Comment (nymmrx) [reply brownie/network/multicall.py:52]: I think `eth_call` code override should be optional as some providers don't support them (e.g. infura)

**Comment by iamdefinitelyahuman** [comment]
Handled in #1125

**Pull request closed without merging**